File: apps/api/src/application/ports/order-repository.ts
import type { Order, OrderCustomer, OrderSummary } from '@/domain/order/order'

export interface OrderItemRequest {
  readonly productId: string
  readonly quantity: number
}

export interface OrderDraft {
  readonly customer: OrderCustomer
  readonly cardLast4: string
  readonly items: readonly OrderItemRequest[]
}

export type Settlement =
  | { readonly outcome: 'approved'; readonly reference: string }
  | { readonly outcome: 'declined'; readonly reason: string }

export interface OrderRepository {
  reserve(draft: OrderDraft): Promise<Order>
  settle(orderId: string, settlement: Settlement): Promise<Order>
  findById(id: string): Promise<Order | null>
  findAll(): Promise<OrderSummary[]>
}

export const ORDER_REPOSITORY = Symbol('OrderRepository')

File: apps/api/src/ecommerce_api/application/ports/order_repository.py
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal, Protocol

from ecommerce_api.domain.order import Order, OrderCustomer, OrderSummary


@dataclass(frozen=True, slots=True)
class OrderItemRequest:
    product_id: str
    quantity: int


@dataclass(frozen=True, slots=True)
class OrderDraft:
    customer: OrderCustomer
    card_last4: str
    items: Sequence[OrderItemRequest]


@dataclass(frozen=True, slots=True)
class Approved:
    reference: str
    outcome: Literal['approved'] = 'approved'


@dataclass(frozen=True, slots=True)
class Declined:
    reason: str
    outcome: Literal['declined'] = 'declined'


Settlement = Approved | Declined


class OrderRepository(Protocol):
    async def reserve(self, draft: OrderDraft) -> Order: ...

    async def settle(self, order_id: str, settlement: Settlement) -> Order: ...

    async def find_by_id(self, id: str) -> Order | None: ...

    async def find_all(self) -> list[OrderSummary]: ...
